# **-01-Basics-of-Python-Programming** 

© by A.Richter (under construction)

<img src=88x31.png  width="10%" style="float: left;"> 

**Preface:** "...A programming language gives you the power to express some ideas, while 
limiting your abilities to express others.“  [LARRY CUBA, computer-animation artis](https://en.wikipedia.org/wiki/Larry_Cuba)

This statement independent of a programming language points to the expression of ideas and creativity. The computer today is used as a tool in the hands of engineers, scientists or any other creative person. Just like a craftsman uses a screwdriver todays knowledgeworker use a specific programming language. But as we all know the toolbox of a craftsman accomodates much more tools than only screwdrivers.

These notebooks are collections of usefull tools including explanations of theory ranging from computerscience to earthsciences, mathmatics and physics. The start is an introduction to the python language. Every notebook has a preface describing roughly the content and the pupose of the discussed topics.

# Overview of the notebook:
    
   1. [Basic Libray Imports, variables and printing](#basiclibraryimports)
   2. [The Zen of Python](#zen)
   3. [Intro to bits and bytes](#bits)
   4. [A little bit of math in Python](#math)
   5. [Python datatypes Part 1](#datatypes)
   6. [If-Else Decision Constructions](#ifthen)
   7. [Loop Constructions](#loops)
   8. [Python datatypes Part 2](#datatypes2)
   9. [Folder and File operations](#folders)
   10. [Functions and methods](#functions)
   11. [Basics of nummeric python (numpy), scientific python (scipy) and TensorFLow](#numpybasics)
   12. [Revision of statistical tools](#stats)
   13. [Preprocessing in Data Analysis](#preprocessing)

**Sources: **
www.introtopython.org  and    http://www.scipy-lectures.org/   


But before we start with learning Python

you should know at least a bit how to use the Commandline **CMD** for WINDOWS or **Terminal** for LINUX & MAC

https://www.thomas-krenn.com/de/wiki/Cmd-Befehle_unter_Windows


1) How do start the Commandline/Terminal in Windows, Linux, macOS? 

2) How do you change your position in the filesystem to a specific folder?

3) How do you create a new txt file in that folder?

4) How do you start Jupyter Notebook in this folder? 


===================================================================

## 1. Basic Libray Imports, variables and printing <a name="basiclibraryimports" ></a>

In general Python is build modular, which means for special things look for the right module and simply import it...

Basic Python Libraries or modules can be imported like this:

In [7]:
import sys

The keyword **'import'** followed by the name of the library you want to use.

There are some specials hidden in the python libraries, for example here:

https://wiki.python.org/moin/UsefulModules                           # Overview

https://www.glowscript.org/#/user/GlowScriptDemos/folder/Examples/   # 3D Examples

https://pypi.org/project/geopy/                                      # Geo tools 

When is a special module needed???

For example see the equation  $10 \cdot 10$ = $10^2$ = 100

That implies that the logarithm to base 10 of 100 must be equal to 2.

So how can I calculate this in python?


In [8]:
#log10(100)

this error message means the base python module library does not know the function log10()

But the python module **math** has a definition of log10() and is therefor needed to calculate log10(100).

So at first I have to import the **math** module

In [9]:
import math

and than I can use it in this way:

In [10]:
math.log10(100)

2.0

A modular function is always called in the form: "**moduleName.functionName(parameters)**" 

### Variables in Python

In [11]:
x=3

Python is very humble concerning variables. There is no need for prior definitions or declaration of the datatype (like in C/C++). Variable are simply assigned to values and can be used immediately.

**The Command "whos" is displaying the current use of variables in the main memory(RAM)**


In [12]:
whos

Variable   Type      Data/Info
------------------------------
math       module    <module 'math' from '/ana<...>h.cpython-36m-darwin.so'>
sys        module    <module 'sys' (built-in)>
x          int       3


### Print out something in python

Printing of an expression or a variable is easily done in this way:

In [13]:
print("HELLO WORLD!")
print("-------------")
print("+-/* Sum: {} ".format(100))

HELLO WORLD!
-------------
+-/* Sum: 100 


Keep in mind that the older versions of python (e.g. Python 2.7) was not using the brackets. Python 3 is very strict in this case and forcing the use of **print("Text2print", variabes)** 

In [14]:
print("The number assigned to the Variable x is",x)

The number assigned to the Variable x is 3


## The Zen of Python <a name="zen" ></a>


In [15]:
import this
#import codecs
#print(this.s.decode('rot13'))
print(this.s)
print(70*'=')

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
Gur Mra bs Clguba, ol Gvz Crgref

Ornhgvshy vf orggre guna htyl.
Rkcyvpvg vf orggre guna vzcyvpvg.
Fvzcyr vf orggre guna pbzcyrk.
Pbzcyrk vf or

In [16]:
d = {}
for c in (65, 97):
    for i in range(26):
        d[chr(i+c)] = chr((i+13) % 26 + c)

print("".join([d.get(c, c) for c in this.s]))

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Intro to Bits and Bytes <a name="bits" ></a>


Computer Science is based on the idea of binary states, whether it's **Light: on/off** or **electr. Voltage: high/low** 


BIT = binary digit – englisch | in german "binäre Ziffer" or  "Binärziffer"

one bit can have two conditions $0$ or $1$ and can be expressed in a vector like this

1bit = $\begin{bmatrix}
0          \\
1        \end{bmatrix}$ 


On a computer numbers can be used in form of the binary system or in form of the decimal system.

In [17]:
print('binary equal decimal')
print('110001   =   ', 1*(2**5)+1*(2**4)+0*(2**3)+0*(2**2)+0*(2**1)+1*(2**0))

binary equal decimal
110001   =    49


But how to calculate this:

In [18]:
x=1*(2**5)+1*(2**4)+0*(2**3)+0*(2**2)+0*(2**1)+1*(2**0)
print(x)

49


or

In [19]:
print(int('110001', 2))

49


1 **Byte** = 8 Bits
= $\begin{bmatrix}
0  & 0&0&0&0&0&0&0       \\
1  & 1&1&1&1&1&1&1     \end{bmatrix}$

so in general 8 options to decide for 1 or 0


In [20]:
print("10010101")
print(int('10010101', 2))

10010101
149


This means that the binary number: 10010101 represents the decimal number: 49

**Gigabytes vs. Gigibytes**

There are different aproaches of defining bigger bunches of information.
For example are **1000 Bytes** called a **kilobyte** which is based on the well known **decimal scale**.

$1000^2$ Bytes = 1MB

$1000^3$ Bytes = 1GB


In [21]:
print(1000**3, "Bytes = 1 GB = gigabyte")

1000000000 Bytes = 1 GB = gigabyte


On the other hand using a binary scale with

1024**3  (GiB)  gibibyte

https://en.wikipedia.org/wiki/Gigabyte

try it yourself...

In [22]:
print(1024**3, "Bytes = 1 GiB = Gigibyte")

1073741824 Bytes = 1 GiB = Gigibyte


## A little bit of math in Python <a name="math" ></a>

Check the Value of a Variable - compare it to another one

In [23]:
x=3

This means that the memory assigned to the variable x is filled with the value 3

In [24]:
x==5

False

The double equal sign (==) is a completly different operation. It implies the comparison of the value stored in the variable x with the value 5. So in other words is 3 equal to 5? No!

In [25]:
x==3

True

But 3 is equal to 3. That's right!

The basic mathmatical operations are expressed for example like this:

In [26]:
x=3
y=5
print(x+y)
print(x-y)
print(x*y)
print(x/y)

8
-2
15
0.6


Division of values in Python 3 is different compared to Python 2 --> Python-2 is using "Int" Python-3 always converts to "Float" before any calculation is done!!!

In [27]:
5/3

1.6666666666666667

In [28]:
5.0/3.0

1.6666666666666667

In [29]:
5**3

125

calculates the the power: $5^3 =5 * 5 * 5=125$


## Python Datatypes <a name="datatypes" ></a>

We will now have a look at basic Datentypes like integer ("**int**") and floating point numbers like "**float**" or "**double**" and of course character types like "**string**".

### Typ: Sting

In [30]:
import sys
S="D:\BHT-BERLIN\nDaten\datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))

D:\BHT-BERLIN
Daten\datei1.shp
<class 'str'>
Size: 79 Bytes


oh look what happened here? The String S was cut in the middle...but why???

the folder  **"\nDaten"**  has a problem, the **"\n"** is interpreteed as a line break!!! 

try ist yourself with a different folder path of your choice...

...better use it this way: " / " instead of " \ " to separate folders

In [31]:
S="D:/BHT-BERLIN/nDaten/datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))


D:/BHT-BERLIN/nDaten/datei1.shp
<class 'str'>
Size: 80 Bytes


or use a so called raw string format **r"... \ ... \ ..."**

In [32]:
S=r"D:\BHT-BERLIN\nDaten\datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))

D:\BHT-BERLIN\nDaten\datei1.shp
<class 'str'>
Size: 80 Bytes


which will work in the same way

In [33]:
y="BHT"
z="Berlin"
print(y+" - "+z)
print(15*"-")
print(y) 
print("Size: {} Bytes".format(sys.getsizeof(y)))
print(z) 
print("Size: {} Bytes".format(sys.getsizeof(z)))

BHT - Berlin
---------------
BHT
Size: 52 Bytes
Berlin
Size: 55 Bytes


Strings can be separated or splitted into sub-strings just by calling the character index. For example the index [0:14] give this sub-string:

In [34]:
print(S)
print(40*'-')
print(S[0:14])

D:\BHT-BERLIN\nDaten\datei1.shp
----------------------------------------
D:\BHT-BERLIN\


A very useful trick in playing with indices is that python supports negativ values:

In [35]:
print(S)
print(40*'-')
print(S[0:-4])

D:\BHT-BERLIN\nDaten\datei1.shp
----------------------------------------
D:\BHT-BERLIN\nDaten\datei1


which can be used to get a filename without any file extension

and to get just the file extension without knowing how long the path to the file is:

In [36]:
print(S)
print(40*'-')
print(S[-4:])

D:\BHT-BERLIN\nDaten\datei1.shp
----------------------------------------
.shp


### Type: Integer 

In [37]:
I=0
print(type(I))
print("Size:{} Bytes".format(sys.getsizeof(I)) )

<class 'int'>
Size:24 Bytes


In [38]:
I=1
print(type(I))
print("Size:{} Bytes".format(sys.getsizeof(I)) )

<class 'int'>
Size:28 Bytes


In [39]:
print("Size:{} Bytes".format(sys.getsizeof(2 ** 63)) )

Size:36 Bytes


### Type: Floating point

In [40]:
F=2.0
#x=6.45678987665454433345666
print(type(F))
print("Size:{} Bytes".format(sys.getsizeof(F)))

<class 'float'>
Size:24 Bytes


In [41]:
F=6.45678987665454433345666
print(type(F))
print("Size:{} Bytes".format(sys.getsizeof(F)))

<class 'float'>
Size:24 Bytes


==> Why are the sizes of INT and FLOAT different?

https://stackoverflow.com/questions/10365624/sys-getsizeofint-returns-an-unreasonably-large-value

https://stackoverflow.com/questions/29826523/python-float-precision-float

## If-Else Decision Constructions <a name="ifthen" ></a>

In Python this is used in this way: 

In [42]:
x=5

if x<0:
    print(x)
    print(2*x)
else:
    print(100-x)

95


Keep an eye on the block that is executed when the if condition (here $x<0$) was evaluated successfully. All statements have to start in the same column (greater than column 1) otherwise python will rise an error.

In [43]:
x=-5

if x<0:
    print(x)
     print(2*x)
else:
    print(100-x)

IndentationError: unexpected indent (<ipython-input-43-e92e48d70095>, line 5)

In [ ]:
x=-5

if x<0:
    print(x)
    print(2*x)
else:
    print(100-x)

this is also mandatory for the **else** block but obviously we have here only on line of code.

Other programming languages like C or Java require brackets to start or to end a block construktion. Python just interpretes all statements in the same column to be part of the current block. 

Regarding the condition that is evaluated in the first if-statement all sorts of logic examples like **...and...**,  **...or...**, **not...**, **...greater than...** constructions can be used:


   - Equals: a == b
   - Not Equals: a != b
   - Less than: a < b
   - Less than or equal to: a <= b
   - Greater than: a > b
   - Greater than or equal to: a >= b
   - ...

Different cases can also be expressed with the **elif** branch: 

In [ ]:
a = 33
b = 33
if b > a:
    print("b is greater than a")
elif a == b:
    print("a and b are equal")

## Loop Constructions <a name="loops" ></a>

Loops are the second basic pillar used in programming languages next to conditionals.

There are different loops available in python, but the most widely used loop is the **for-loop** 

### while-Loop

This is a simple example for a while-loop construction:

In [ ]:
x=0
while x<10:
    print(x)
    x=x+1

The while-loop is mainly helpful in cases of permanently running codes. Accordingly we can choose a condition which is always true so the codes in the while-block are repeated until an external break signal is sent. [Embedded codes](#https://micropython.org/) for example make use of that.  

### for-loop

The power of the for-loop construction is mainly that it is implemented as a **for-each algorithm**.
This means that the user does not need to know how many item there are to be iterarted over before the loop starts. The for loop works on python lists that inherrently know their start, end and their number of items in between.

In [ ]:
for i in range(35,40):
    print(i)

In [ ]:
for i in range(35,55,5):
    print(i)

The function **range(START,STOP,STEP)** provides a highly flexible tool in combination with the for-Loop

Be aware that the specified STOP Value is always evaluated at the index before the value itself.


References: 

   - http://introtopython.org/while_input.html
   - https://www.geeksforgeeks.org/loops-in-python/

## Python Datatypes Part 2 <a name="datatypes2" ></a>

### Type: list

A python list can be created by asigning a name to the squared brackets just like in the following line: 

In [ ]:
usernames = []

In [ ]:
whos

This is an empty list-object. We can check the type of constructed object by the command **whos**.

A List can also be created already filled with values like this example shows. The for-loop as shown in the previous section can deal with list objects in that way: 

In [ ]:
names = ['peter', 'paul', 'jenny']

for name in names:
    print("Hello, " + name.title() + "!") name.

The List-object has inherrited a lot of methods from its super-class. Just type **"name."** + TAB key to see all the options:

In [ ]:
name.

In [ ]:
names.

For example the **append()** method can be used to add an entry to the list.

In [ ]:
names.append("julia")

In [ ]:
for name in names:
    print("Hello, " + name.title() + "!")

In [ ]:
for name in names:
    print(name)

The List values are enumerated thats why we can do something like that:

In [ ]:
for index, name in enumerate(names):
    print("Index: " + str(index) + " Name: " + name.title())

Lists can also be created with numeric values instead of strings:

In [ ]:
numbers = [1, 3, 4, 2, 7, 9, 24, 8, 5, 11]
print(numbers)

In [ ]:
numbers.sort()
print(numbers)

In [ ]:
numbers.sort(reverse=True)
print(numbers)

In [ ]:
number_count = len(numbers)
print(number_count)

### Type: tuple

Comparing the type **tuple** to the type **list** we can easily see a difference in the used brackets. Tuples are generated via the curved **()** brackets as parentheses and not with squared brackets.

Tuples are basically list that cannot be changed. As we have seen before in a list elements can be added or deleted from it. This is not possible with tuples as there are no such methods to append or delete in that class.

In [ ]:
colors = ('red', 'green', 'blue')
print("The first color is: " + colors[0])

print("\nThe available colors are:")
for color in colors:
    print("- " + color)

In [ ]:
colors.append("yellow")

### Type: dictionary

is a list of object saved in a key - value pair structure.

## Folder and File Operations <a name="folders" ></a>

To Search a Folder for a specific File Type, the following construction can be used.

glob is a module to search folderpaths and make lists of the contents.

In [ ]:
import glob
Folder='figures/'
#Folder='../../_Data_ori_Tif/AOI2_small_6_Test/'
# https://docs.python.org/2/library/glob.html 
# http://stackoverflow.com/questions/3207219/how-to-list-all-files-of-a-directory-in-python

JPGs=glob.glob(Folder+"*.jpg")
PNGs=glob.glob(Folder+"*.png")

for j in JPGs:
    print(j)

print(30*'-') 
    
for p in PNGs:
    print(p)

In this little example the **for-each** character is shown best. All we need to do is generating a python list befor using it in the for-loop construction. Data can be easily manipulated this way.

## Functions and methods <a name="functions" ></a>

Functions are a great and important topic in Python. We have already used some of them in all the examples before like **print(..)** or **math.log10(...)**. Those are so called build-in functions which can be used right from scratch or after import of specific module. 

Python is build modular and open so that every user can write his own functions just as needed to solve the specific requirements

Of course python supports the **Object-Oriented-Programming paradigm (OOP)**


   - Method is called by its name, but it is associated to an object (dependent).
   - A method is implicitly passed the object on which it is invoked.
   - It may or may not return any data.
   - A method can operate on the data (instance variables) that is contained by the corresponding class


To summerize the functions:
    
    
    
   - Function is block of code that is also called by its name. (independent)
   - The function can have different parameters or may not have any at all. If any data (parameters) are passed, they are passed explicitly.
   - It may or may not return any data.
   - Function does not deal with Class and its instance concept.



   - Simply, function and method both look similar as they perform in almost similar way, but the key difference is the concept of ‘Class and its Object‘.
   - Functions can be called only by its name, as it is defined independently. But methods can’t be called by its name only, we need to invoke the class by a reference of that class in which it is defined, i.e. method is defined within a class and hence they are dependent on that class.


Simply write the literal **"def"** at the beginning of a line
followed by the name of the function e.g. **"MyFunct()"** (a list of input parameter is separeted by comma) than the colon to start the function-block. All statements need to be aligned to the same column just like in case of conditionals or loops. The end of a function is marked by the literal **return** follow by the variable(s) that is given back. 

In [ ]:
def MyFunct(af,bf):
    cf=af+bf
    return cf

a=4
b=4
print("a =",a,"| b =",b)
print(2*"-------")

print("Summe(a,b): {}".format(MyFunct(a,b)))

In [ ]:
# Functions: indent blocks with 4 spaces
def hello_world(name, times=3):
    print("Say",name*times)
    return 

hello_world("yes")

A function to compute the exponential $f(x)=e^x$ by a taylor approximation:

$$e^x= \sum_{n=0}^{\infty} \frac{x^n}{n!} =1+x+\frac{x^2}{2!}+\frac{x^3}{3!}+\frac{x^4}{4!}+...$$

In [3]:
import math
def taylorexp(inp, acc):
    sum=0
    for i in range(0,acc):
        sum=sum+(inp**i)/math.factorial(i)
    return sum

In [4]:
math.exp(4)

54.598150033144236

In [5]:
print(taylorexp(4,100))
print(taylorexp(4,100) - math.exp(4))

54.598150033144265
2.842170943040401e-14


#### on methods...

For example the use of the build-in functions from the **math module** refers to the class definitions and uses the "functions" call in a object-oriented way:  **modulename.functionname(parameter)** in case of the build-in functions I will call them **methods** from now on to use a clear language.

In [ ]:
import math
math.log10(100)

In [ ]:
# Python 3 User-Defined Method 
class ABC : 
    def method_abc (self): 
        print("I am in method_abc of ABC class. ") 

class_ref = ABC() # object of ABC class 
class_ref.method_abc() 


In [1]:
class TaylorTools:
    
    def taylorexp(inp, acc):
        import math
        sum=0
        for i in range(0,acc):
            sum=sum+(inp**i)/math.factorial(i)
        return sum

In [2]:
TaylorTools.taylorexp(4,100)

54.598150033144265

References:
    
   - https://www.geeksforgeeks.org/difference-method-function-python/
   - http://introtopython.org/classes.html

## Basics of Nummeric Python (numpy), Scientific Python (scipy) and TensorFlow <a name="numpybasics" ></a>


Many scientific problems can be described with algebraic expressions in form of a [**tensor formalism**](#https://en.wikipedia.org/wiki/Tensor). Very general speaking we have used tensors in the examples above and also in daily life without knowing it maybe. Tensors are divided into classes according to their dimensionality:

   - tensor 0. order: **Scalar**, e.g. a single real number like **42**
   - tensor 1. order: **Vector**, e.g. a row or column array of numbers $a_i=[a_1 a_2 a_3]$
   - tensor 2. order: **Matrix**, e.g. a 2-D array of numbers $A_{i,j}$
   - tensor 3. order: **3D-Tensor**, e.g. a 3-D array of numbers aranged like a cube $A_{i,j,k}$
   - tensor 4. order: **4D-Tensor**, e.g. a vector of 3D-Tensors (Cubes)  $A_{i,j,k,n}$
   - tensor 5. order: **5D-Tensor**, e.g. a matrix of 3D-Tensors (Cubes)  $A_{i,j,k,n,l}$


[Numpy](#https://www.numpy.org/) and [Scipy](#https://www.scipy.org/) as well as [Googles Tensorflow](#https://www.tensorflow.org/) make use of this approach.

Data structures like that can be found in daily life for example:

 - any kind of table has the character of a matrix or sec. order tensor
 - a layer of grids in a GIS-Analysis can be seen as a 3D-Tensor
 - or an imagefile with two or more channels can be a 3D-Tensor
 - when we look at series of Images (each one with multiple channels) we have a 4D-Tensor data structure
 - video streams are seqeunces of images over time which can be organized as a 5D-Tensor

<img src="https://hackernoon.com/hn-images/1*_D5ZvufDS38WkhK9rK32hQ.jpeg"  width="70%" style="float: left;"> 

References: 
  - https://hackernoon.com/learning-ai-if-you-suck-at-math-p4-tensors-illustrated-with-cats-27f0002c9b32
  - https://simplyml.com/linear-algebra-shootout-numpy-vs-theano-vs-tensorflow-2/

In numeric python the basic datastructure that we will use is the **array**. Here we need to distinguish between the used dimensions of arrays. A numpy array is in general implemented as a multi-dimensional array (ndarray) but has a special subclass of called **matrix objects** which are strictly 2-dimensional and therefor they inherit all the attributes and methods of ndarrays. If $a$ and $b$ are defined as numpy matrices of the same dimensions, then the product $a*b$ is their matrix product. When $a$ and $b$ are defined as numpy arrays, then their product $a*b$ is the element-wise product of their values.

The exact differences between np.array and np.matrix will be discussed in the examples below. To start we will have a look at the basic methods **np.arange()**, **np.zeros()** and **np.ones()** which can be used to create array structures from scratch.

In [44]:
import numpy as np

### Creating arrays

We can easily create a **Vector** or **a 1d array** in diffenrent ways. For example the method **np.arange(a,b)** generates a vector with 64bit integer values between the numbers a and b. 

In [45]:
e=np.arange(4,11)
print(e)
print("Dimensions: ", e.shape)
print("Datatype: ",e.dtype)

[ 4  5  6  7  8  9 10]
Dimensions:  (7,)
Datatype:  int64


In [46]:
e

array([ 4,  5,  6,  7,  8,  9, 10])

It is an array structure with dimension 1

**np.arange(START,STOP,STEP)** is as flexible as the regular range() function used in [section 7 (for loops](#loops)) to use different step sizes.

In [47]:
f=np.arange(1,8.5,0.5)
print(f)

[ 1.   1.5  2.   2.5  3.   3.5  4.   4.5  5.   5.5  6.   6.5  7.   7.5  8. ]


The method **np.ones(dim,dim)** can be used in the same way to create vectors or multi-dimensional arrays (matrix) with **equal values of 1** .

In [48]:
x=np.ones((3,1))
y=np.ones((1,3))
print("x: \n",x)
print(20*'-') 
print("y: ",y)

x: 
 [[ 1.]
 [ 1.]
 [ 1.]]
--------------------
y:  [[ 1.  1.  1.]]


We can see here that depending on the choice of the dimensions np.ones((3,1)) or np.ones((1,3)) a column or a row of values is generated.

In the same way arrays with values of 0 can be created by the method **np.zeros(dim,dim)**

In [49]:
np.zeros((5,5))

array([[ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.]])

In [50]:
print(np.ones((5,5)))

[[ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]]


In [51]:
np.ones((3,3))

array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.]])

just note: The print-out of the arrays hides the dtype.

If more than 2 indices are given in np.zeros() than a multidimensional array is generated.

In [52]:
np.ones((3,3,3))

array([[[ 1.,  1.,  1.],
        [ 1.,  1.,  1.],
        [ 1.,  1.,  1.]],

       [[ 1.,  1.,  1.],
        [ 1.,  1.,  1.],
        [ 1.,  1.,  1.]],

       [[ 1.,  1.,  1.],
        [ 1.,  1.,  1.],
        [ 1.,  1.,  1.]]])

In [53]:
np.ones((3,3,3,3))

array([[[[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]]],


       [[[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]]],


       [[[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]],

        [[ 1.,  1.,  1.],
         [ 1.,  1.,  1.],
         [ 1.,  1.,  1.]]]])

### Manipulating arrays

A simple 4x4 array can also be initilized manually in this way:

In [54]:
A=np.array([[2,4,3,7],[1,2,3,4],[-2,4,5,7],[4,3,7,9]])
print("A: \n",A)

A: 
 [[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]


The transposed array (col and row exchanged) is called with the method **".T"** 

In [55]:
print("Transpose(A): \n",A.T)

Transpose(A): 
 [[ 2  1 -2  4]
 [ 4  2  4  3]
 [ 3  3  5  7]
 [ 7  4  7  9]]


Arrays can be manipulated or data can be extracted out of the structures. For example we can calculate the sum of each column or each row by the following statements:

In [56]:
b=np.sum(A, axis=0)
print("b: ",b)

b:  [ 5 13 18 27]


... or the sum of each row

In [57]:
c=np.sum(A, axis=1)
print("c: ",c)

c:  [16 10 14 23]


... the parameter **axis=0/1** is specifying the sum:  axis=0 for columns and axis=1 for rows

We can pick an element from the array by its index. Note that all indexing in Python starts with 0.

In [58]:
d=A[2,0]
print("A: \n",A)
print("  ") 
print('The Value at index [row=2,col=0] is: {} '.format(d))

print('Size of the Value (by .itemsize): {} Bytes '.format(d.itemsize))
print('Size of the Value (by .nbytes): {} Bytes '.format(d.nbytes))

A: 
 [[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]
  
The Value at index [row=2,col=0] is: -2 
Size of the Value (by .itemsize): 8 Bytes 
Size of the Value (by .nbytes): 8 Bytes 


We can also pick a complete row from the array and play with the index in different ways:

In [59]:
print("A: \n",A)
print('---------------------')
print('Row:1 by Col:0 to end := all Cols A[1,:]')
print('-------------')
print(A[1,:])
print(' ')
print('All Cols with specific Index A[0:4]')
print('-------------')
print(A[1,0:4])
print(' ')
print('All without the 1st Col, A[1,1:4] ')
print('-------------')
print(A[1,1:4])
print('-------------')

A: 
 [[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]
---------------------
Row:1 by Col:0 to end := all Cols A[1,:]
-------------
[1 2 3 4]
 
All Cols with specific Index A[0:4]
-------------
[1 2 3 4]
 
All without the 1st Col, A[1,1:4] 
-------------
[2 3 4]
-------------


Now we want to change the arrays values by **replacing the second row**

In [60]:
print(A) 
A2=A
print(2*'---------')
A2[1,:]=[5,6,7,8]

print(A2)

[[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]
------------------
[[ 2  4  3  7]
 [ 5  6  7  8]
 [-2  4  5  7]
 [ 4  3  7  9]]


### The Matrix product of arrays

In case of numpy arrays the matrix product is calculated by a method called **np.dot()**. 

In [61]:
print(np.dot(A,A2))

[[ 46  65  98 130]
 [ 58 108 148 204]
 [ 34  57  96 116]
 [ 45  89 131 182]]


When we use the normal multiplication operator $*$ instead of np.dot() an element-wise multiplication of the two arrays is calculated (presupposed both have the same dimensions).

In [62]:
print(A*A2)

[[ 4 16  9 49]
 [25 36 49 64]
 [ 4 16 25 49]
 [16  9 49 81]]


As already said above the general numpy array structure has the subclass numpy matrix which is defined for the special case of 2-dimensional arrays. A numpy matrix is created by the method **np.mat()** as shown in the following example.

In [63]:
import numpy as np

a=np.mat('4 3; 2 1')
b=np.mat('1 2; 3 4')
print("a: \n",a)
print(10*'=')
print("b: \n",b)
print(10*'=')
print("a*b: \n",a*b)

a: 
 [[4 3]
 [2 1]]
b: 
 [[1 2]
 [3 4]]
a*b: 
 [[13 20]
 [ 5  8]]


When we use the np.mat() method the matrix product is calculated by the $*$ operator. The transposed matrix is displayed analog to the array by the **".T"** operator.

In [64]:
print("Transpose(a): \n",a.T) # Transpose
print(10*'=')
print("conjugate Transpose(a): \n",a.H) # conjugate Transpose in case of complex arrays
print(10*'=')
print("Inverse(a): \n",a.I) # Inverse

Transpose(a): 
 [[4 2]
 [3 1]]
conjugate Transpose(a): 
 [[4 2]
 [3 1]]
Inverse(a): 
 [[-0.5  1.5]
 [ 1.  -2. ]]


#### In Python >= 3.5 the @ operator is working to compute the matrix product of arrays

In [22]:
a=np.array([[4, 3], [2, 1]])
b=np.array([[1, 2], [3, 4]])
print(a@b)

[[13 20]
 [ 5  8]]


In [23]:
print("a: \n",a) 
print(10*'=')
print("Transpose(a): \n",a.T)
print(10*'=')
print("b: \n",b)
print(10*'=')
print("a*b: \n",np.dot(a,b))
print(10*'=')
print("Inverse(a): \n",np.linalg.inv(a))

a: 
 [[4 3]
 [2 1]]
Transpose(a): 
 [[4 2]
 [3 1]]
b: 
 [[1 2]
 [3 4]]
a*b: 
 [[13 20]
 [ 5  8]]
Inverse(a): 
 [[-0.5  1.5]
 [ 1.  -2. ]]


### The Identity Matrix (I or E) Symmetric and Orthogonal Matrixes

In [20]:
import numpy as np

I=np.mat('1. 0.; 0. 1.')
print(I)

[[1. 0.]
 [0. 1.]]


In [3]:
print(np.identity(3))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [4]:
print(np.eye(3))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


**Matrix diagonals** and **determinant**

Quadratic matrixes have a main diagonal from the upper left corner to the lower right corner. In case of the identity matrix this is where the ones are sitting. The opposit is the secondary diagonal from the lower left corner to the upper right corner. The main diagonal is the mirroring axis for the transpose operation (exchange of rows and cols... see above) 

In [31]:
a = np.array([[1, 2], [3, 4]])
print(np.linalg.det(a))

-2.0000000000000004


The **trace of a matrix** is the sum of the elements on the main diagonal.

In [5]:
import numpy as np
print(np.eye(3))
print(10*"=")
print("trace of Matrix:",np.trace(np.eye(3)))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
trace of Matrix: 3.0


**Symmetric matrix** are independent of the transpose operation which means $A^T=A$


In [21]:
import numpy as np

b = np.random.randint(-100,100,size=(3,3))
print(b)
print(20*"=")
b_symm = (b + b.T)/2

print(b_symm)
print(20*"=")
print(b_symm.T)

[[-10  42  65]
 [-80  38  67]
 [-47  54 -86]]
[[-10.  -19.    9. ]
 [-19.   38.   60.5]
 [  9.   60.5 -86. ]]
[[-10.  -19.    9. ]
 [-19.   38.   60.5]
 [  9.   60.5 -86. ]]


An **Orthogonal matrix** are a special kind of matrix where the transpose is equal to the inverse matrix $A^T=A^{-1}$


so the matrix product $A \times  A^T = I $ and results in the identity matrix $I$

In [30]:
a=1/5*np.array([[3, 4], [-4, 3]])
print(a)
print(20*'=')
ainv=np.linalg.inv(a)
print("Inverse(a): \n",ainv)
print(20*'=')
print("Transpose(a): \n",a.T)
print(20*'=')
print("a x a.T\n",np.round(a@a.T))

[[ 0.6  0.8]
 [-0.8  0.6]]
Inverse(a): 
 [[ 0.6 -0.8]
 [ 0.8  0.6]]
Transpose(a): 
 [[ 0.6 -0.8]
 [ 0.8  0.6]]
a x a.T
 [[1. 0.]
 [0. 1.]]


### Scientific Python (scipy)
SciPy contains additional routines needed in scientific work: for example, routines for computing integrals numerically, solving differential equations, optimization, and sparse matrices.

In [ ]:
from scipy.spatial import ConvexHull
import matplotlib.pyplot as plt

points = np.random.rand(30, 2) # 30 random points in 2-D
plt.plot(points[:,0], points[:,1], 'o')
plt.show()

hull = ConvexHull(points)
plt.plot(points[:,0], points[:,1], 'o')
for simplex in hull.simplices:
    plt.plot(points[simplex,0], points[simplex,1], 'k-')
plt.show()

In [ ]:
from scipy.interpolate import interp1d

x  = np.linspace(0, 10, num=11, endpoint=True)
y  = np.cos(-x**2/9.0)
f  = interp1d(x, y)
f2 = interp1d(x, y, kind='cubic')

xnew = np.linspace(0, 10, num=41, endpoint=True)
plt.plot(x, y, 'o', xnew, f(xnew), '-', xnew, f2(xnew), '--')
plt.legend(['data', 'linear', 'cubic'], loc='best')
plt.show()

### TensorFlow

pip install tensorflow==2.0.0-beta1 

#### On TensorFlow and KERAS


xyz

In [2]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)


Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 3s 58us/sample - loss: 0.3001 - accuracy: 0.9122
Epoch 2/5
60000/60000 [==============================] - 3s 55us/sample - loss: 0.1460 - accuracy: 0.9553
Epoch 3/5
60000/60000 [==============================] - 3s 54us/sample - loss: 0.1082 - accuracy: 0.9669
Epoch 4/5
60000/60000 [==============================] - 3s 56us/sample - loss: 0.0874 - accuracy: 0.9731
Epoch 5/5
10000/10000 [==============================] - 0s 35us/sample - loss: 0.0762 - accuracy: 0.9759s - loss: 0.0868 - accuracy: 0.


[0.07623121768273414, 0.9759]

References: 

 - https://stackoverflow.com/questions/4151128/what-are-the-differences-between-numpy-arrays-and-matrices-which-one-should-i-u
 
 - http://scipy-lectures.org/index.html
 - https://www.tutorialspoint.com/scipy
 - https://docs.scipy.org/doc/scipy/reference/tutorial/index.html

## Revision of statistical tools <a name="stats" ></a>

Statistical methods are a working field with a long history of over 200 years. The great variety of branches ranges from bayesian analysis over optimization to parts of machine learning but here we will only focus on a few tools with general application in multiple scientific diciplines: Mean and standard-deviation. Later on in the following notebooks we will get back to more sophisticated statistpts.

### Mean Value and Standard-deviation

The term standard deviation was first used[17] in writing by Karl Pearson[18] in 1894, following his use of it in lectures. This was as a replacement for earlier alternative names for the same idea: for example, Gauss used mean error.[19]


We start with random values, compute their mean and accordingly the standard deviation.


We can generate a list of 20 random values,
around 50 with the numpy method **np.random.randn(20,1)**

In [ ]:
x=50+10*np.random.randn(20,1)
print(x)

#### Calculting the Mean of the values

according to the formal definition: $ \bar{X} =\frac{\sum\nolimits_{i=1}^{20} x_i}{20}$ 

In [ ]:
Xmean=np.sum(x)/20

print(Xmean)

or calculate it with a numpy method **np.mean()**:

In [ ]:
np.mean(x)

try it  yourself:

#### Calculating the Standard Deviation of the values

according to the formal definition: $ std = \sqrt{\frac{\sum\nolimits_{i=1}^{20}(x_i - \bar{X})^2}{20-1}}$ where $\bar{X}$ is the Mean value of x

In [ ]:
np.sqrt( np.sum((x-np.mean(x))**2)/19 ) 

In [ ]:
np.std(x)

**!!! NOT** the same value, but how is numpy calculating here? Be carefull when you use this method !!! 

Read the documentation first...!!! (RTFM :-)

In [ ]:
print(np.std(x,ddof=1) )

The parameter **ddof : int, optional** means "**D**elta **D**egrees **o**f **F**reedom". The divisor used in calculations is **(N - ddof)**, where N represents the number of elements. **By default ddof is zero.**


In [ ]:
np.sqrt( np.sum((x-np.mean(x))**2)/20 ) 

This is value form before calculated with the defaults of np.std()

To explain this with a very simple dataset we can have a look at this.

In [ ]:
numbers=np.array([[6,7,8,9,10]])
print(numbers)

In [ ]:
np.mean(numbers)

In [ ]:
np.std(numbers,ddof=1)

is called the standard deviation, with (n-1) degrees of freedom. 

**But why do we have to use (n-1) instead of n???**

The reasoning in the natural or engineering sciences is this: 

If we assume every random value in this dataset represents a meassurement we can say that every independent meassurement is equal to an independent information.

In the formal definition of the mean value (cp. the calculation above) we have n independent information available and spent one information to calculate the mean. The mean value is more likely to be near to the true value (that we try to meassure) than every single meassurement is.

In the formal definition of the standard deviation (cp. the calculation above) the mean value is used and this is the reason why for this calculation there are no more n information available but still (n-1) independent information. For every further calculation of more sophisticated statistical parameters (e.g. coefficient of variation ) we spent in the same way an independent information.

In economics for example this reasoning is different and the default values (ddof=0) is used.

In [ ]:
print("Standard deviation : ",np.std(numbers))

print("Variance           : ",np.var(numbers))

$s^2=\frac{(6-8)^2+(7-8)^2+(8-8)^2+(9-8)^2+(10-8)^2 }{5}=\frac{4+1+0+1+4 }{5} =\frac{10 }{5}=2$

by the way this is called a **Maximum Likelihood estimation** of the standard deviation. More later.

References: 

 - https://www.frustfrei-lernen.de/mathematik/standardabweichung-berechnen.html
 - https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy
 - https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.std.html
 - https://en.wikipedia.org/wiki/Standard_deviation

## Preprocessing in Data Analysis <a name="preprocessing" ></a>

Most data in its raw form is not adequate for analysis methods as shown above.

A dataset can have single points of no data, points of infinity or the values can simply be in the wrong range. 



- get number and index of noData Points (NaN)
print('Anzahl NaN in Dataset-Name: {}'.format(np.sum(np.isnan(../Path-to-Dataset)*1)))


- get number and index of (infinity positions)
print('...these positions are set to Zero(0) :')
Dataset[np.isinf(Dataset)]=0


:Mean od Median insection

In [ ]:
#Anzeigen der Bilder und Histogramme

from scripts.BasisFunctions import showImage

    
    
for layer in sorted(LayerTrain):
    print layer
    showImage( LayerTrain.get(layer))
    #print Layers.get(layer)

In [ ]:
#Prüfung auf Datenvollständigkeit
#NaN - Not a Number
# Infty....

print 'Anzahl infty in TIF_EucDist_faults: {}'.format(np.sum(np.isinf(TIF_r_EucDist_faults)*1))

TIF_r_EucDist_roads[np.isinf(TIF_r_EucDist_roads)]=0
TIF_r_EucDist_takhar[np.isinf(TIF_r_EucDist_takhar)]=0
TIF_r_DistStreams[np.isinf(TIF_r_DistStreams)]=0

print 'Anzahl infty in TIF_EucDist_faults: {}'.format(np.sum(np.isinf(TIF_r_EucDist_faults)*1))
print 'Anzahl infty in TIF_EucDist_roads: {}'.format(np.sum(np.isinf(TIF_r_EucDist_roads)*1))
print 'Anzahl infty in TIF_EucDist_takhar: {}'.format(np.sum(np.isinf(TIF_r_EucDist_takhar)*1))
print 'Anzahl infty in TIF_DistStreams: {}'.format(np.sum(np.isinf(TIF_r_DistStreams)*1))


Change the scaling of the values in case they are unballanced.


In [ ]:

print('for example apply a logarithmic scale to Distance Rasters')

TIF_r_EucDist_faults =np.log(TIF_r_EucDist_faults)


Change Datatype of Rasters in case they do not match to the rest of the data

In [ ]:

TIF_r_precip=np.float32(skio.imread(Folder+'TIF_precip.tif'))
print 'TIF_r_precip Shape: ', TIF_r_precip.shape,TIF_r_precip.dtype
Layers['TIF_r_precip']=TIF_r_precip


Create a Python dictionary to work efficintly with a bunch of raster arrays

In [ ]:
Layers=dict()
Layers['TIF_r_dem'] = TIF_r_dem
Layers['TIF_r_precip']=TIF_r_precip

Generate/Split off new layers from specific numbers of a single array or reorganize the layers

In [ ]:
Layers['TIF_n_LanduseRAL']=((Layers.get('TIF_n_lc93_32') ==1) | (Layers.get('TIF_n_lc93_32') ==5))*1.0
print 'TIF_n_LanduseRAL ', Layers['TIF_n_LanduseRAL'].shape
        
Layers['TIF_n_LanduseRL']=(Layers.get('TIF_n_lc93_32') ==2)*1.0
print 'TIF_n_LanduseRL ', Layers['TIF_n_LanduseRL'].shape
        
Layers['TIF_n_LanduseIRA']=((Layers.get('TIF_n_lc93_32') ==4) | (Layers.get('TIF_n_lc93_32') ==9) |(Layers.get('TIF_n_lc93_32') ==10))*1.0
print 'TIF_n_LanduseIRA' ,Layers['TIF_n_LanduseIRA'].shape


In [ ]:
Layers['TIF_n_VOLC']= (Layers.get('TIF_n_lith_type') == 1)*1.0  # % Volcanic ...... Vulkanisches Gestein
print 'TIF_n_VOLC ', Layers['TIF_n_VOLC'].shape
        
Layers['TIF_n_SEDI']= (Layers.get('TIF_n_lith_type') == 2)*1.0  # % Sedimentary ... Sedimente
print 'TIF_n_SEDI ', Layers['TIF_n_SEDI'].shape
       
Layers['TIF_n_PLUT']= (Layers.get('TIF_n_lith_type') == 3)*1.0  # % Plutonic....... Plutonisches Gestein
print 'TIF_n_PLUT ', Layers['TIF_n_PLUT'].shape


Reshaping Arrays and Vectors 

In the following notebook **-01a-Optimizing-your-SourceCode** I will show some usefull construction and techniques that ensure a better performance than the classical programming paradigm.
